### DISCUSSION TITLE
June 2023 impressions of the project

### DISCUSSION THREAD

**Discussion by SvdSinner** [ant-design-blazor/ant-design-blazor#3295] [General]
I just finished reviewing Blazor UI frameworks for a major application my company is about to embark on, and I want to share my results for the benefit of this project.

This project (version 0.15.0) was one of two finalist out of the 12 frameworks we looked at, but ultimately it lost out to [Radzen-Blazor](https://github.com/radzenhq/radzen-blazor) as our go forward framework.  I spent 3 weeks creating a demo of how AntBlazor could work with the project and another senior developer did the same with [Radzen-Blazor](https://github.com/radzenhq/radzen-blazor)

It beat every closed source framework we evaluated (Telerik, Infragistics and SyncFusion).  The advantage of having the source code when trying to figure out an advanced feature was huge.  I honestly don't know how closed source components could compete against this massive advantage of being open-source.

It beat out 7 open source projects: (Skclusive.Material.Component, Blazored, MatBlazor, MudBlazor, Blazorize, BlazorStrap and PanoramicData.Blazor)  It had a full pallet of options both for general information display/editing, and charts and graphics to visualize data at an executive level.  We were also very attracted to the Mentions component that was a great fit for one of the features we plan to implement.  The Skeleton feature was also impressive and a big plus for us.

In the final round, some big positives and some negatives showed themselves.  I'm passing these on as feedback to help your community.  

**Pros:** 

- The components work great when used together.  On several occasions, I initially tried to blend regular HTML with AntBlazor components and my regular html parts ended up looking funny by default.  Each time this happened, I switched my raw HTML to the AntBlazor equivalent (E.G.  Used Space and SpaceItem tags instead of raw Div, Span and P tags) and the AntBlazor defaults looked substantially better than the defaults of my raw HTML.  Using AntBlazor components over their raw HTML equivalents dropped the time required to use CSS styling to get a descent look by a huge amount.
- The Mentions control has no equivalent on any of the other systems we looked at.  Great free functionality for tagging people in comments, etc.
- The TimePicker and DatePicker worked well together for setting a single DateTime variable.
- Add-ons like Badge and Tooltip made sprinkling in these features easy and painless and didn't break any functionality we tried it with.  (E.G.  The Segmented control had no problems with us wrapping it's SegmentedItem tags inside of Tooltip tags.)
- The controls were a complete toolset.  Most every display metaphor we wanted had controls that displayed them nicely.
- The Ant-Design-Charts-Blazor projects is a spectacular add-on that opened up numerous visual display options to us.  This was a big reason that this project made it to the final round.

**Neutral Comments:**

- The Space tag is weirdly named.  I initially thought it was for displaying chunks of blank space.  Very useful once I figured it out it's true purpose.
- Great selection of decent but not stellar available icons
- The Chinese text used in the demo project was not useful to non-Chinese speakers.  (I wished there was a way to translate the source code from inside Visual Studio.)   Browsers handled the translation easily when the demo was run, so it wasn't a major issue.

**Cons:**

- I could never get the demo project for Ant-Design-Blazor to run properly on my local machine.  It would run, but displayed like it couldn't load its CSS.  The console displayed numerous errors (Primarily related to JSInterop), and I suspect it requires some additional undocumented installation step before it could work.   Additionally, https://localhost:5001/_content/AntDesign.Docs/css/default.css and https://localhost:5001/_content/AntDesign/js/ant-design-blazor.js caused 404 errors.  (<-- probably why the page looked like it had broken CSS and why JSInterop threw errors)   I was still able to use the page at antblazor.com, and could find answers in the source code, but it was a big disappointment that the demo app was not runnable.
- The demo project from Ant-Design-Charts-Blazor worked locally, but the page on [AntBlazor.com](https://antblazor.com/en-US/charts/introduce) was completely obsolete and had outdated examples that did not work at all with the current source code.
- The documentation of both projects was poor.  It was tough to figure out how to modify things and use templates.  If I didn't have access to the source code, I would have never been able to do things like using a custom TitleTemplate on a Card.
- The control set lacks a basic ToggleButton control.   This was a glaring omission in most every framework we evaluated.  It was bizarre that custom coding was required to make a basic click-star-to-favorite-click-star-again-to-unfavorite control.
- The charts documentation was terrible, with very little explanation of what options were available and what they did.  The config and data editor concept in the demos of the current source code was great, but was unusable due to lack of documentation on what format the config and data needed to be in.
- The Mentions control is an awesome idea but was so poorly documented that I couldn't make it work in the time I had.
- The Modal control throws exceptions when opened.  The errors were all related to JSInterop.  This may have been related to why I could never run the demo project on my local machine.  (Is there an un-documented setup requirement?)

**Summary:**  This framework is a very awesome diamond in the rough.  It was the framework to beat going into the final round of evaluation.  However, severe problems with the documentation removed it from consideration.  The fact that the demo project wouldn't run on any of the machines I tried to install it on, was shocking.  The outdated charting documentation on the main web-site was shocking since the newer version was already checked in and working.

**Comment by ElderJames** [comment]
Thank you very much @SvdSinner ! None of us expected this to work out, even though you didn't choose us in the end.

We really appreciate your thoughtful comments.

We think what you said is right, our current documentation is not very detailed and friendly. We put most of our effort into development, and less into documentation, except where it is necessary to improve.

For building the source code on local machine, we have been introduce the way in https://github.com/ant-design-blazor/ant-design-blazor#local, please try.

We know there's a lot more to be done, but we're still a fully open source project with no commercial support behind it, and we'd love to get more people involved. We don't have enough contributors to get involved at the moment.

**Reply by ElderJames** [reply]
For the toggle button, please check out the checkable Tag https://antblazor.com/components/tag#components-tag-demo-Checkable

**Reply by SvdSinner** [reply]
> None of us expected this to work out, even though you didn't choose us in the end.

Don't sell yourself short.  You guys were a very close second and put a lot of distance between this project and third place.

> We put most of our effort into development, and less into documentation, except where it is necessary to improve.

It truly shows. The high level of the development was why you stood out from 10 out of 11 other frameworks we looked at.  Seriously, it clear there are great developers contributing.  

> For building the source code on local machine, we have been introduce the way in https://github.com/ant-design-blazor/ant-design-blazor#local, please try.

I had already followed this once, but tried redoing the npm install, build and npm start steps again.  This actually improved things and fixed the broken CSS, but still give a 404 on https://localhost:5001/_framework/blazor.polyfill.min.js and https://localhost:5001/_content/AntDesign.Charts/g2plot.js.map.  Initially, none of the links would run and several jsinterop errors were thrown.  However, I shut down the process and restarted it and it seems to be working except for the 2 404 errors mentioned above.  I might suggest that someone debugs these steps on a blank machine and update the instructions.

Additionally, there should be instructions on how Visual Studio users can run the project from F5 rather than via the console. (I.E.  Which of the various start-up projects should I be choosing)  That way if there are issues, one isn't left wondering if the issues are because of selecting the wrong start-up project.    Note:  I had a similar result with running through F5 as I did with using NPM start:  After re-running the npm install, the site did not even load through F5.  However, killing it and trying F5 a second time worked just fine.   Possibly add a notice that sometimes the server doesn't always build correctly the first time but restarting it often helps?

As a final note of encouragement, I really don't think the documentation needs THAT much work.  

- Improve the instructions for local install (Make a clean machine image and confirm the instructions work)
- Add a "Development concepts" page in the documentation about general concepts like how Templates work, what types of objects can be a DataSource, etc.
- Flesh out the Mentions documentation.  That is an awesome control but the documentation examples are mostly "TODO"
- Either update the Ant-Design-Charts-Blazor docs for the newest version or clearly indicate what version of the nuget package is required for the examples in the docs to work, and clearly mention that the later versions aren't fully ready yet.

I really think that with a few minor tweaks this can be the leading Blazor UI framework.
